language/vm/vm-runtime/src/chain_state.rs

```diff
@@ -22,24 +22,44 @@ use vm_runtime_types::{loaded_data::struct_def::StructDef, values::GlobalValue};
 
 /// Trait that describes what Move bytecode runtime expects from the Libra blockchain.
 pub trait ChainState {
+    // ---
     // Gas operations
+    // ---
+
     fn deduct_gas(&mut self, amount: GasUnits<GasCarrier>) -> VMResult<()>;
     fn remaining_gas(&self) -> GasUnits<GasCarrier>;
 
-    // StateStore operations. Ideally the api should look like:
-    // fn read_data(&self, ap: &AccessPath) -> VMResult<Vec<u8>>;
-    // fn write_data(&mut self, ap: &AccessPath, data: Vec<u8>) -> VMResult<()>;
-    // However this is not implementable due to the current implementation of MoveVM: data are
-    // organized as a tree of GlobalRefs.
-    /// Get a mutable reference to a resource stored on chain.
-    fn load_data(
+    // ---
+    // StateStore operations
+    // ---
+
+    // An alternative for these APIs might look like:
+    //
+    //   fn read_data(&self, ap: &AccessPath) -> VMResult<Vec<u8>>;
+    //   fn write_data(&mut self, ap: &AccessPath, data: Vec<u8>) -> VMResult<()>;
+    //
+    // However, this would make the Move VM responsible for deserialization -- in particular,
+    // caching deserialized results leads to a big performance improvement. But this directly
+    // conflicts with the goal of the Move VM to be as stateless as possible. Hence the burden of
+    // deserialization (and caching) is placed on the implementer of this trait.
+
+    /// Get the serialized format of a `CompiledModule` from chain given a `ModuleId`.
+    fn load_module(&self, module: &ModuleId) -> VMResult<Vec<u8>>;
+
+    /// Get a reference to a resource stored on chain.
+    fn borrow_resource(
         &mut self,
         ap: &AccessPath,
         def: StructDef,
-    ) -> VMResult<&mut Option<(StructDef, GlobalValue)>>;
+    ) -> VMResult<Option<&GlobalValue>>;
 
-    /// Get the serialized format of a `CompiledModule` from chain given a `ModuleId`.
-    fn load_module(&self, module: &ModuleId) -> VMResult<Vec<u8>>;
+    /// Get ownership of a resource stored on chain, and mark it to be deleted at the end of this
```

Comment:
"at the end of this transaction" sounds a bit weird to me. Should we put something like transfer the ownership of a resource from on chain to local stack?

Reply:
Yeah, changed. thanks!